grades/api.py

```diff
@@ -177,11 +177,14 @@ def freeze_user_final_grade(user, course_run, raise_on_exception=False):
                     course_run.edx_course_key
                 )
             ) from ex
-    return FinalGrade.objects.create(
+    # the final grade at this point should not exists, but putting a `get_or_create`
+    # should solve the problem when the function is called synchronously from the dashboard REST API multiple times
+    final_grade_obj, _ = FinalGrade.objects.get_or_create(
```

Comment:
I'm not sure if we should be looking up the `FinalGrade` with all of these parameters. The unique constraint is on `user` and `course_run`, so if the problem was that the record already existed in the database, we will still get the unique constraint error. Maybe we want an exception in that case however? I'm not sure